analyses/org.eclipse.tracecompass.incubator.inandout.core/src/org/eclipse/tracecompass/incubator/internal/inandout/core/analysis/InAndOutDataProviderFactory.java

```diff
@@ -0,0 +1,244 @@
+    private static final String NAME = "In And Out Analysis"; //$NON-NLS-1$
+    private static final String DESCRIPTION = "Configure In And Out analysis using file description"; //$NON-NLS-1$
+
+    private static final Table<String, ITmfTrace, ITmfConfiguration> fTmfConfigurationTable = HashBasedTable.create();
+
+    static {
+        Bundle bundle = Platform.getBundle(Activator.PLUGIN_ID);
+        IPath defaultPath = new Path(SegmentSpecifierConfiguration.IN_AND_OUT_CONFIG_SOURCE_SCHEMA);
+        URL url = FileLocator.find(bundle, defaultPath, null);
+        File schemaFile = null;
+        try {
+            schemaFile = new File(FileLocator.toFileURL(url).toURI());
+        } catch (URISyntaxException | IOException e) {
+            Activator.getInstance().logError("Failed to read schema file: " + SegmentSpecifierConfiguration.IN_AND_OUT_CONFIG_SOURCE_SCHEMA, e); //$NON-NLS-1$
+        }
+        CONFIG_SOURCE_TYPE = new TmfConfigurationSourceType.Builder()
+                .setId(SegmentSpecifierConfiguration.IN_AND_OUT_CONFIG_SOURCE_TYPE_ID)
+                .setDescription(DESCRIPTION)
+                .setName(NAME)
+                .setSchemaFile(schemaFile)
+                .build();
+    }
+
+    private static final IDataProviderDescriptor DESCRIPTOR = new DataProviderDescriptor.Builder()
+            .setId(ID)
+            .setName("InAndOut Configurator") //$NON-NLS-1$
+            .setDescription("Configure custom InAndOut analysis") //$NON-NLS-1$
+            .setProviderType(ProviderType.NONE)
+            .build();
+
+    /**
+     * Default constructor
+     */
+    public InAndOutDataProviderFactory() {
+        TmfSignalManager.register(this);
+    }
+
+    @Override
+    public @Nullable ITmfTreeDataProvider<? extends ITmfTreeDataModel> createProvider(ITmfTrace trace) {
+        return null;
+    }
+
+    @SuppressWarnings("null")
+    @Override
+    public Collection<IDataProviderDescriptor> getDescriptors(ITmfTrace trace) {
+        return List.of(DESCRIPTOR);
+    }
+
+    @SuppressWarnings("null")
+    @Override
+    public @NonNull List<@NonNull ITmfConfigurationSourceType> getConfigurationSourceTypes() {
+        return List.of(CONFIG_SOURCE_TYPE);
+    }
+
+    @SuppressWarnings( "null")
+    @Override
+    public @NonNull List<@NonNull IDataProviderDescriptor> createDataProviderDescriptors(String typeId, ITmfTrace trace, String jsonParameters) throws TmfConfigurationException {
+        /*
+         * Note only one config type is available for this.
+         */
+        ITmfConfiguration defaultConfig = new TmfConfiguration.Builder()
+                .setId("Dummy ID") // not used //$NON-NLS-1$
+                .setSourceTypeId(typeId)
+                .setName("InAndOut config") //$NON-NLS-1$
+                .setDescription("Custom In And Out").build(); //$NON-NLS-1$
+
+        ITmfConfiguration config = TmfJsonConfiguration.fromJsonString(jsonParameters, defaultConfig);
```

Comment:
My intuition would be to define the mandatory/common configuration properties (id, name, description) outside so that they could all be described in the TSP, while leaving the parameters object as completely opaque and only defined by the specific extension in its schema's 'properties' element.

Reply:
I had that intuition too initially, some implication on the implementation on how the input json is generated in the client and how it can be validated in the server (using a validation library) If we put it outside we will need to get a separate JSON Object to feed the json validator. If the name/description is part of the schema, that JSON object has to include it. If not, the schema can't have it. Both solutions are valid.

Reply:
I'm ok with both solutions as long as it's clear in TSP what needs to be included in the map.

How about if the map contains 'name', 'id', 'description' and 'properties' where the value of properties is an object that matches the properties in the JSON schema? I think it would be less risky than having the application-specific elements at the same root level as the mandatory configuration elements, to avoid key name conflict.